components/server/ee/src/workspace/workspace-factory.ts

```diff
@@ -373,15 +373,22 @@ export class WorkspaceFactoryEE extends WorkspaceFactory {
             }
 
             if (OpenPrebuildContext.is(context.originalContext)) {
+                if (CommitContext.is(buildWorkspace.context)) {
+                    if (
+                        CommitContext.is(context.originalContext) &&
+                        CommitContext.computeHash(context.originalContext) !==
+                            CommitContext.computeHash(buildWorkspace.context)
+                    ) {
+                        // If the current context has a newer/different commit hash than the prebuild
```

Comment:
Thanks for updating this comment! At first glance, this change looks like it could "fix" the situation when prebuild commit == branch HEAD.

While this is definitely an improvement over the status quo, it does leave the cases with prebuild commit != branch HEAD with an awkward non-branch checkout.

I wonder if a (follow-up) more general fix could be possible, e.g. by making the workspace Git initializer respect _both_ `revision` _and_ `branch`, such that you could check out a branch @ an earlier HEAD. 🤔 But this would probably need further discussion, and thus warrant a follow-up issue.

Meanwhile, let's wait for the (correct) CI build to finish so that we can confirm & merge this fix. 🙌

Reply:
> While this is definitely an improvement over the status quo, it does leave the cases with prebuild commit != branch HEAD with an awkward non-branch checkout.

I think this makes sense though. As an example: in my project I'm using prebuilds to compile my scala codebase so that when the workspace opens I can run my project without having to compile anything. It saves me around 10 minutes per workspace. If I were to try running a workspace on an older prebuild but the latest code was checked out instead, my process to run my scala project would automatically start compiling everything.

I guess this brings up an interesting question though: what's the reasoning for allowing old prebuilds to be used at all? What's the use case?

Reply:
Also, I don't think it's possible to checkout a branch at an earlier commit. The branch is just a reference to a single commit. I suppose it would be possible to reset the branch (i.e. like `git reset <commit> --hard`).

Like you mentioned though, that's all probably a bigger discussion.

Reply:
> I guess this brings up an interesting question though: what's the reasoning for allowing old prebuilds to be used at all? What's the use case?

What you mentioned before this about saving 10 minutes by starting from a Prebuild that is already finished is exactly the use case. This can be a life saver, e.g. for projects with very long build times, or with a high frequency of new commits being merged all the time.

FYI, there is another feature that is upcoming: A `Use Last Successful Prebuild` button right next to the `Skip Prebuild` button in the `Prebuild in Progress` screen. This would also allow you to save a lot of time when you start a workspace while a prebuild is still running.

> Also, I don't think it's possible to checkout a branch at an earlier commit. The branch is just a reference to a single commit. I suppose it would be possible to reset the branch (i.e. like `git reset <commit> --hard`).

Yes, I was implying that you'd end up with a local branch, that has the same name as the upstream branch, but is hard-reset to the specified commit.

I.e. it would not be in "detached HEAD" state, as is currently the case when you open a Prebuild directly, but instead it would keep the desired branch name (just at an earlier commit).

Instead of showing:

<img width="524" alt="Screenshot 2022-10-21 at 10 44 55" src="https://user-images.githubusercontent.com/599268/197153764-5e60e760-cea9-4486-9484-7b50a900fc7d.png">

It would show:

<img width="552" alt="Screenshot 2022-10-21 at 10 45 14" src="https://user-images.githubusercontent.com/599268/197153768-517c55d2-9150-45f6-8ebf-83c17f2c1e22.png">

And `git status` would say `Your branch is behind 'origin/fix_prebuild_open' by 1 commit`